internal/fs/sd_windows.go

```diff
@@ -0,0 +1,431 @@
+
+// Flags for restore without admin permissions. If there are no admin permissions, only the DACL from the SD can be restored and owner and group will be set based on the current user.
+var lowRestoreSecurityFlags windows.SECURITY_INFORMATION = windows.DACL_SECURITY_INFORMATION | windows.ATTRIBUTE_SECURITY_INFORMATION | windows.PROTECTED_DACL_SECURITY_INFORMATION
+
+// GetSecurityDescriptor takes the path of the file and returns the SecurityDescriptor for the file.
+// This needs admin permissions or SeBackupPrivilege for getting the full SD.
+// If there are no admin permissions, only the current user's owner, group and DACL will be got.
+func GetSecurityDescriptor(filePath string) (securityDescriptor *[]byte, err error) {
+	onceBackup.Do(enableBackupPrivilege)
+
+	var sd *windows.SECURITY_DESCRIPTOR
+
+	if lowerPrivileges {
+		sd, err = getNamedSecurityInfoLow(filePath)
+	} else {
+		sd, err = getNamedSecurityInfoHigh(filePath)
+	}
+	if err != nil {
+		if !lowerPrivileges && isHandlePrivilegeNotHeldError(err) {
+			lowerPrivileges = true
+			sd, err = getNamedSecurityInfoLow(filePath)
+			if err != nil {
+				return nil, fmt.Errorf("get low-level named security info failed with: %w", err)
+			}
+		} else {
+			return nil, fmt.Errorf("get named security info failed with: %w", err)
+		}
+	}
+
+	sdBytes, err := securityDescriptorStructToBytes(sd)
+	if err != nil {
+		return nil, fmt.Errorf("convert security descriptor to bytes failed: %w", err)
+	}
+	return &sdBytes, nil
+}
+
+// SetSecurityDescriptor sets the SecurityDescriptor for the file at the specified path.
+// This needs admin permissions or SeRestorePrivilege, SeSecurityPrivilege and SeTakeOwnershipPrivilege
+// for setting the full SD.
+// If there are no admin permissions/required privileges, only the DACL from the SD can be set and
+// owner and group will be set based on the current user.
+func SetSecurityDescriptor(filePath string, securityDescriptor *[]byte) error {
+	onceRestore.Do(enableRestorePrivilege)
+	// Set the security descriptor on the file
+	sd, err := SecurityDescriptorBytesToStruct(*securityDescriptor)
+	if err != nil {
+		return fmt.Errorf("error converting bytes to security descriptor: %w", err)
+	}
+
+	owner, _, err := sd.Owner()
+	if err != nil {
+		//Do not set partial values.
+		owner = nil
+	}
+	group, _, err := sd.Group()
+	if err != nil {
+		//Do not set partial values.
+		group = nil
+	}
+	dacl, _, err := sd.DACL()
+	if err != nil {
+		//Do not set partial values.
+		dacl = nil
+	}
+	sacl, _, err := sd.SACL()
+	if err != nil {
+		//Do not set partial values.
+		sacl = nil
+	}
+
+	if lowerPrivileges {
+		err = setNamedSecurityInfoLow(filePath, dacl)
+	} else {
+		err = setNamedSecurityInfoHigh(filePath, owner, group, dacl, sacl)
+	}
+
+	if err != nil {
+		if isHandlePrivilegeNotHeldError(err) {
```

Comment:
Should also check for `!lowerPrivileges`

Reply:
Fixed